python/lsst/daf/butler/gen2convert/translators.py

```diff
@@ -0,0 +1,275 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import itertools
+from abc import ABCMeta, abstractmethod
+
+__all__ = ("Translator", "NoSkyMapError", "KeyHandler", "CopyKeyHandler", "ConstantKeyHandler")
+
+
+class NoSkyMapError(LookupError):
+    """Exception thrown when a Translator cannot be created because it needs
+    a nonexistent SkyMap.
+    """
+    pass
+
+
+class KeyHandler(metaclass=ABCMeta):
+    """Base class for Translator helpers that each handle just one Gen3 Data ID key.
+
+    Parameters
+    ----------
+    gen3key : `str`
+        Name of the Gen3 Data ID key (DataUnit link field name) populated by this
+        handler.
+    gen3unit : `str`
+        Name of the Gen3 DataUnit associated with `gen3key`.
+    """
+
+    __slots__ = ("gen3key", "gen3unit")
+
+    def __init__(self, gen3key, gen3unit):
+        self.gen3key = gen3key
+        self.gen3unit = gen3unit
+
+    def translate(self, gen2id, gen3id, skyMap, skyMapName):
+        gen3id[self.gen3key] = self.extract(gen2id, skyMap=skyMap, skyMapName=skyMapName)
+
+    @abstractmethod
+    def extract(self, gen2id, skyMap, skyMapName):
+        raise NotImplementedError()
+
+
+class Translator:
+    """Callable object that translates Gen2 Data IDs to Gen3 Data IDs for a
+    particular DatasetType.
+
+    Translators should usually be constructed via the `makeMatching` method.
+
+    Parameters
+    ----------
+    handlers : `list`
+        A list of KeyHandlers this Translator should use.
+    skyMap : `BaseSkyMap`
+        SkyMap instance used to define any Tract or Patch DataUnits.
+    skyMapName : `str`
+        Gen3 SkyMap DataUnit name to be associated with any Tract or Patch
+        DataUnits.
```

Comment:
Maybe it should be, in the future, but `BaseSkyMap` (and its subclasses) are not new, and I'm a little wary of adding a `name` attribute since it can't be derived from the SkyMap's actual contents.